tests.py

```diff
@@ -1,129 +1,523 @@
+                    return msg
+
+            time.sleep(0.05)
+        # end while
+
+        raise Exception(
+            f'Didn`t receive "{_type}" in time; have: {[type(m).__name__ for m in self.msgs]}'
+        )
+
+    def stop(self):
+        if self.lsp_client.is_initialized:
+            self.lsp_client.shutdown()  # send shutdown...
+            self.get_msg_by_type(lsp.Shutdown)  # receive shutdown...
+            self.lsp_client.exit()  # send exit...
+            self._process_qs()  # give data to send-thread
+        else:
+            self.process.kill()
+
+
+test_langservers = pathlib.Path(__name__).absolute().parent / "test_langservers"
+
+
+SERVER_PYLS = "pyls"
+SERVER_JS = "js"
+SERVER_CLANGD_10 = "clangd_10"
+SERVER_CLANGD_11 = "clangd_11"
+SERVER_GOPLS = "gopls"
+
+SERVER_COMMANDS = {
+    SERVER_PYLS: lambda: [sys.executable, "-m", "pyls"],
+    SERVER_JS: lambda: [
+        test_langservers / "node_modules/.bin/javascript-typescript-stdio"
+    ],
+    SERVER_CLANGD_10: lambda: [
+        next(test_langservers.glob("clangd_10.*")) / "bin" / "clangd"
+    ],
+    SERVER_CLANGD_11: lambda: [
+        next(test_langservers.glob("clangd_11.*")) / "bin" / "clangd"
+    ],
+    SERVER_GOPLS: lambda: ["gopls"],
+}
+
+
+def start_server(server_name, tmp_path_factory):
+    command = SERVER_COMMANDS[server_name]
+    command = command()
+    project_root = tmp_path_factory.mktemp("tmp_" + server_name)
+
+    if server_name == SERVER_GOPLS:
+        # create file(s) before starting server, jic
+        for fn, text in files_go.items():
+            path = project_root / fn
+            path.write_text(text)
+
     with subprocess.Popen(
         command, stdin=subprocess.PIPE, stdout=subprocess.PIPE
     ) as process:
-        lsp_client = lsp.Client(trace="verbose", root_uri=project_root.as_uri())
-
-        def make_event_iterator():
-            while True:
-                process.stdin.write(lsp_client.send())
-                process.stdin.flush()
-                # TODO: multiple bytes not work
-                received = process.stdout.read(1)
-                assert received
-                yield from lsp_client.recv(received)
-
-        yield (lsp_client, make_event_iterator())
-
-
-def do_stuff_with_a_langserver(tmp_path, filename, file_content, language_id, command):
-    path = tmp_path / filename
-    path.write_text(file_content)
-
-    with run_langserver(tmp_path, command) as (lsp_client, event_iter):
-        inited = next(event_iter)
-        assert isinstance(inited, lsp.Initialized)
-        lsp_client.did_open(
-            lsp.TextDocumentItem(
-                uri=path.as_uri(), languageId=language_id, text=file_content, version=0
+        tserver = ThreadedServer(process, project_root.as_uri())
+
+        yield (tserver, project_root)
+
+        if tserver.msgs:
+            print(
+                "* unprocessed messages:",
+                ", ".join(type(m).__name__ for m in tserver.msgs),
             )
```

Comment:
When are there messages left? Can we `assert not tserver.msgs`?

Reply:
That would be messages that were not requested (like logs). It's nice to see when test failed.

> Can we assert not `tserver.msgs`?

Not consumed messages do not indicate an error, so we should not